libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.cpp

```diff
@@ -858,19 +858,74 @@ bool HTTPClient::connect(void)
 
     _client->setTimeout(_tcpTimeout);
 
-    if(!_client->connect(_host.c_str(), _port)) {
-        DEBUG_HTTPCLIENT("[HTTP-Client] failed connect to %s:%u\n", _host.c_str(), _port);
+    String host = _host;
+    uint16_t port = _port;
+    if (_useProxy)
+    {
+        host = _proxyHost;
+        port = _proxyPort;
+        DEBUG_HTTPCLIENT("[HTTP-Client] using proxy %s:%u\n", host.c_str(), port);
+    }
+
+    if(!_client->connect(host.c_str(), port)) {
+        DEBUG_HTTPCLIENT("[HTTP-Client] failed connect to %s:%u\n", host.c_str(), port);
         return false;
     }
 
-    DEBUG_HTTPCLIENT("[HTTP-Client] connected to %s:%u\n", _host.c_str(), _port);
+    DEBUG_HTTPCLIENT("[HTTP-Client] connected to %s:%u\n", host.c_str(), port);
 
 #ifdef ESP8266
     _client->setNoDelay(true);
 #endif
     return connected();
 }
 
+bool HTTPClient::setProxyHost(const String &proxyUrl)
```

Comment:
most of this is copied from beginInternal. Would it be worthwhile to pull the shared logic out into a URI parser?

Reply:
I gave it a shot in this commit https://github.com/esp8266/Arduino/pull/8094/commits/178495725b3db7bf3f1cf24f6965255b3e556b5c